src/components/DynamicForm/components/DynamicFormContainer.jsx

```diff
@@ -21,58 +21,99 @@ import { isFieldInvalid, isEmpty } from "./utilities";
  */
 class DynamicFormContainer extends React.Component {
   state = {
-    disabled: true,
-    form_data: {},
-    questions: []
+    form_data: {}, // { field_name: user response(s) }
+    questions: [], // detailed in prop types
+    disabled: true, // overall DF Container submit control
+    field_errors: {}, // individual field errors
   }
 
   componentDidMount() {
     const { initialData, purpose, questions } = this.props;
-    const state = this._initializeState(purpose, questions);
+    const state = { questions };
 
-    const default_form_data = state.form_data;
+    const persistence = window.localStorage.getItem(purpose);
+    if (persistence) {
+      const persisted_state = this._getStateFromPersistence(state, persistence, initialData);
+      return this.setState(persisted_state);
+    }
+    
+    // get initial default values
+    state.form_data = this._getDefaultFormData(questions);
+    // merge with initialData if available
+    if (initialData) state.form_data = { ...state.form_data, ...initialData };
 
-    // merge with prop form_data initial values
-    if (initialData) state.form_data = { ...default_form_data, ...initialData };
+    // validate all answers (defaults and any provided by initialData)
+    // uses onValidate() or isFieldInvalid() on each question / form_data field value
+    const { disabled, field_errors } = this._validateAllAnswers(state.form_data, questions);
+    state.disabled = disabled;
+    state.field_errors = field_errors;
 
-    // check if the form data (with potential initialData) is valid to submit
-    state.disabled = this._hasEmptyAnswers(state.form_data);
     return this.setState(state);
   }
 
   componentDidUpdate(prevProps) {
     // when the form_data is updated from onFormChange
+    const { form_data, field_errors, disabled } = this.state;
     // or it receives new questions (for multi-question sets)
-    const { form_data, disabled } = this.state;
     const { purpose, persistence, questions } = this.props;
 
     // persistence in LS
     if (persistence) {
-      // only persist disabled state and form_data
-      const persistedData = JSON.stringify({ disabled, form_data });
+      const persistedData = JSON.stringify({ form_data, field_errors, disabled });
       localStorage.setItem(purpose, persistedData);
     }
 
     // update form_data when a new question set is introduced
     // handles cases where multiple question sets may be introduced by
     // the DF Wrapper managing the DF Container
-    if (!isEqual(questions, prevProps.questions)) { // only update if question set changes
+    if (!isEqual(questions, prevProps.questions)) { // only update if question set changes, performance of deep equal?
```

Comment:
What's the other option? isEqual exists on mismatch - good enough

Reply:
well the fact that its performing a deep equality check every single time the component updates. when 99% of the time there is only a single question set (we only have 1 component that introduces variable question sets for the same container).

so we are performing a deep equality on every update to support a < 1% use case. just wondering if the performance impact is noticable.

one alternative would be to pass a prop flag "multi" or something. this could be checked as 

```js
if (this.props.multi && deepEquality(...)) { ... }
```